### PR TITLE
initiated login

### PR DESCRIPTION
**Pull request by meomancer**
Making login dialog and login connection

### CODE REVIEW

File: source/gui/login/cadasta_login.py [hunk 1]
```diff
@@ -22,20 +22,23 @@
...
-        # http://qt-project.org/doc/qt-4.8/designer-using-a-ui-file.html
-        # #widgets-and-dialogs-with-auto-connect
+        super(CadastaLogin, self).__init__(parent)
         self.setupUi(self)
+        self.loginButton.clicked.connect(self.login)
+
+    def login(self):  # real signature unknown; restored from __doc__
+        username = self.usernameInput.displayText()
+        password = self.passwordInput.displayText()
+        print "login with %s/%s" % (username, password)
```
Comment (Gustry) [inline source/gui/login/cadasta_login.py]: We should not use some 'print' statement, especially for password

File: source/mixin/network_mixin.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    def connectionFinished(self):
+        print(self.result_connection)
+
+    @pyqtSlot()
+    def connectionReadData(self):
+        self.result_connection += self.reply.readAll()
+
+    def connect(self):
+        print "connect it"
+        QObject.connect(
```
Comment (Gustry) [inline source/mixin/network_mixin.py]: 'QObject.connect' is the old way to connect signals. You should do something like : https://github.com/inasafe/inasafe/blob/develop/safe/gui/tools/osm_downloader_dialog.py#L99
'python_object.signal_name.connect(function)'

File: source/mixin/network_mixin.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+            self.reply, SIGNAL('finished()'),
+            self, SLOT('connectionFinished()'))
+
+    def connectGET(self):
+        self.reply = self.get(self.req)
+        self.connect()
+
+    def connectPOST(self, data):
+        print "create parameter : %s" % data
```
Comment (Gustry) [inline source/mixin/network_mixin.py]: We should use the 'logging' system for that instead of 'print'

**Review by Gustry — COMMENTED** [review]
I made some small comments

**Comment by meomancer** [comment]
hi @Gustry 
i got blocker for this
do you know why i can't connect? for login i mean? i tried the API outside of dialog, it is working fine, but from dialog when login button clicked, it is not connected.

File: source/api/login.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
+# -*- coding: utf-8 -*-
+import json
+from source.mixin.network_mixin import NetworkMixin
+from PyQt4.QtCore import pyqtSlot, QByteArray
```
Comment (Gustry) [inline source/api/login.py]: Can you change this to my issue #11 ?

File: source/gui/login/cadasta_login.py [hunk 5]
```diff
@@ -22,20 +22,23 @@
 """
 
 import os
-
 from PyQt4 import QtGui, uic
```
Comment (Gustry) [inline source/gui/login/cadasta_login.py]: Can you change this to my issue #11 ?

File: source/mixin/network_mixin.py (hunk 3 above)
Comment (ismailsunni) [reply source/mixin/network_mixin.py]: Example of logging.
```
import logging

LOGGER = logging.getLogger('CadastaQGISPlugin')

LOGGER.info('Info')
LOGGER.debug('Debug')
```

Comment (Gustry) [reply source/mixin/network_mixin.py]: And to add a handler to forward the message to QGIS : https://github.com/inasafe/inasafe/blob/develop/safe/common/custom_logging.py#L43

File: source/mixin/network_mixin.py (hunk 2 above)
Comment (timlinux) [reply source/mixin/network_mixin.py]: @meomancer Also please read carefully https://github.com/inasafe/inasafe/wiki/Coding-Standards#qt-guidelines (and the rest of the coding guidelines in that document) - it should be the basis for all your QGIS Python plugin coding.

File: __init__.py [hunk 6]
```diff
@@ -21,6 +21,10 @@
  ***************************************************************************/
```
Comment (timlinux) [inline __init__.py]: Please see https://github.com/inasafe/inasafe/wiki/Coding-Standards#standard-headers for the expected format of headers. In this case copyright would be to cadasta, not to InaSAFE....

File: source/api/login.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
+# -*- coding: utf-8 -*-
```
Comment (timlinux) [inline source/api/login.py]: As above, use standard header format described here: https://github.com/inasafe/inasafe/wiki/Coding-Standards#standard-headers

File: source/api/login.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+
+class Login(NetworkMixin):
+    # LOGIN class
+    request_url = URL_TARGET + 'account/login/?'
+
+    def __init__(self, USERNAME, PASSWORD, output_label):
+        NetworkMixin.__init__(self)
+        # extract data
+        postData = QByteArray()
```
Comment (timlinux) [inline source/api/login.py]: Please use pythonic variable naming conventions - postData -> post_data etc.

File: source/api/login.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+        # extract result
+        result = self.result_connection.data()
+        result = json.loads(result)
+        # check result
+        if 'auth_token' in result:
+            auth_token = result['auth_token']
+            output_result = "auth_token is %s" % auth_token
+        else:
+            output_result = result
+        self.label_3.setText(output_result)
```
Comment (timlinux) [inline source/api/login.py]: Lets use semantically rich names for widgets : label_3 -> results_label

File: source/gui/login/cadasta_login.py [hunk 10]
```diff
@@ -22,20 +22,22 @@
...
-        # Set up the user interface from Designer.
-        # After setupUI you can access any designer object by doing
-        # self.<objectname>, and you can use autoconnect slots - see
-        # http://qt-project.org/doc/qt-4.8/designer-using-a-ui-file.html
-        # #widgets-and-dialogs-with-auto-connect
+        super(CadastaLogin, self).__init__(parent)
         self.setupUi(self)
+        self.loginButton.clicked.connect(self.login)
+
+    def login(self):  # real signature unknown; restored from __doc__
```
Comment (timlinux) [inline source/gui/login/cadasta_login.py]: Every method should have a doctoring....

File: source/mixin/network_mixin.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
+from PyQt4.QtCore import *
```
Comment (timlinux) [inline source/mixin/network_mixin.py]: Standard header please https://github.com/inasafe/inasafe/wiki/Coding-Standards#standard-headers

File: source/mixin/network_mixin.py [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+from PyQt4.QtNetwork import *
+
+
+class NetworkMixin(QNetworkAccessManager):
```
Comment (timlinux) [inline source/mixin/network_mixin.py]: Every class should have  a doctoring. See https://github.com/inasafe/inasafe/wiki/Coding-Standards#doc-strings-and-comments

Also whenever adding a new class or method you should annotate it with the .. versionadded:: sphinx note - see https://github.com/inasafe/inasafe/wiki/Coding-Standards#annotating-api-changes-and-additions

File: source/mixin/network_mixin.py [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+        print "connect it"
+        QObject.connect(
+            self.reply, SIGNAL('readyRead()'),
+            self, SLOT('connectionReadData()'))
+
+        QObject.connect(
+            self.reply, SIGNAL('finished()'),
+            self, SLOT('connectionFinished()'))
+
+    def connectGET(self):
```
Comment (timlinux) [inline source/mixin/network_mixin.py]: Please use pythonic method names : connectGET -> connect_get. Better to also use semantically meaningful names e.g. setup_get_slot

File: source/mixin/network_mixin.py [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+        self.reply = self.get(self.req)
+        self.connect()
+
+    def connectPOST(self, data):
```
Comment (timlinux) [inline source/mixin/network_mixin.py]: As above connectPost -> setup_post_slot

Comment (timlinux) [reply source/mixin/network_mixin.py]: Also we should probably name these as private members if they are not intended to be part of the public interface for the class (setup_post_slot -> _setup_post_slot).

File: test/test_login.py [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+    def setUp(self):
+        """Runs before each test."""
+        self.username = 'irwan.kartoza'
+        self.password = 'Irwankartoza1!'
+
+    def tearDown(self):
+        """Runs after each test."""
+        self.dialog = None
+
+    def test_login(self):
```
Comment (timlinux) [inline test/test_login.py]: No self.assert here?

File: source/gui/login/cadasta_login.py [hunk 16]
```diff
@@ -21,21 +21,34 @@
...
         """Constructor."""
-        super(CadastaDialog, self).__init__(parent)
-        # Set up the user interface from Designer.
-        # After setupUI you can access any designer object by doing
-        # self.<objectname>, and you can use autoconnect slots - see
-        # http://qt-project.org/doc/qt-4.8/designer-using-a-ui-file.html
-        # #widgets-and-dialogs-with-auto-connect
+        super(CadastaLogin, self).__init__(parent)
         self.setupUi(self)
+        self.loginButton.clicked.connect(self.login)
```
Comment (timlinux) [inline source/gui/login/cadasta_login.py]: Note that I prefer that we use pythonic naming for widgets in ui files too. loginButton -> login_button etc. I know its a bit confusing as Qt naming is not pythonic, but I prefer wherever we can to follow python conventions and keep non pythonic stuff 'at the edge' of our code.

File: source/mixin/network_mixin.py (hunk 13 above)
Comment (timlinux) [reply source/mixin/network_mixin.py]: Also I don't know if it is good to overload the reserved method name 'connect' it might create side effects

Comment (dimasciput) [reply source/mixin/network_mixin.py]: Thanks, I already updated this. https://github.com/kartoza/cadasta-qgis-plugin/pull/10/commits/c9b9955ccfe684229863e936af909086099a5dd1

**Comment by dimasciput** [comment]
Hi @timlinux, @ismailsunni, @Gustry, can you guys review this PR again? We've been updated this PR based on your comments, and I think this PR is getting bigger for an initial PR. Thanks!

**Review by timlinux — COMMENTED** [review]
LGTM

File: cadasta/mixin/network_mixin.py [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+    """A mixin that can be used to send network request and receive replies."""
+
+    request_url = str
+    error = None
+    results = None
+
+    def __init__(self, request_url=None):
+        if request_url is not None:
+            self.request_url = request_url
+        self.manager = QNetworkAccessManager()
```
Comment (Gustry) [inline cadasta/mixin/network_mixin.py]: You should use the QNetworkManager provided by QGIS : https://qgis.org/api/classQgsNetworkAccessManager.html
It's using the settings from QGIS desktop by default.

Comment (timlinux) [reply cadasta/mixin/network_mixin.py]: Yes @dimasciput, thanks @Gustry. This is critical - its the first thing I asked @cchristelis to file in your issue tracker: #1

**Comment by timlinux** [comment]
@dimasciput 

> Hi @timlinux, @ismailsunni, @Gustry, can you guys review this PR again? We've been updated this PR based on your comments, and I think this PR is getting bigger for an initial PR. Thanks!

Yeah lets get it merged ASAP and rather do many small incremental changes. I'd like to see a PR for travis tests happening soon....

**Review by ismailsunni — APPROVED** [review]
LGTM

**Pull request merged**